okhttp/src/main/kotlin/okhttp3/internal/cache/CacheLock.kt

```diff
@@ -0,0 +1,44 @@
+package okhttp3.internal.cache
+
+import android.annotation.SuppressLint
+import java.io.Closeable
+import java.nio.channels.FileChannel
+import java.nio.file.StandardOpenOption
+import java.util.Collections
+import okhttp3.internal.platform.Platform
+import okio.FileSystem
+import okio.Path
+
+internal object CacheLock {
+  private val openCaches = Collections.synchronizedMap(mutableMapOf<Path, Exception>())
+
+  fun openLock(fileSystem: FileSystem, directory: Path): Closeable {
+    return if (fileSystem == FileSystem.SYSTEM && !Platform.isAndroid) {
+      fileSystemLock(directory)
+    } else {
+      inMemoryLock(directory)
+    }
+  }
+
+  @SuppressLint("NewApi")
+  fun inMemoryLock(directory: Path): Closeable {
+    // TODO solution for Android N?
+    val existing = openCaches.putIfAbsent(directory, Exception("CacheLock($directory)"))
```

Comment:
Is `putIfAbsent()` atomic on synchronized maps? I assume it’s not?

Reply:
```
        @Override
        public V putIfAbsent(K key, V value) {
            synchronized (mutex) {return m.putIfAbsent(key, value);}
        }
```

Reply:
Oh neat. Today I learned!